src/util/log.c

```diff
@@ -0,0 +1,648 @@
+         * our message.
+         * Sadly, `%r' is still no standardized format specifier, so we need
+         * the separate calls into log_append() here.
+         */
+
+        log->offset = 0;
+
+        if (format && *format) {
+                log_appendf(log, "<%d>", log->level ?: 6);
+                log_vappendf(log, format, args);
+                log_appends(log, "\n");
+        }
+
+        if (log->error)
+                r = error_trace(log->error);
+        else if (log->offset)
+                r = log_stream_send(log);
+        else
+                r = 0;
+
+        return error_trace(r);
+}
+
+static int log_commit_journal(Log *log, const char *format, va_list args) {
+        int r;
+
+        /*
+         * Lets append `MESSAGE=%s\n' to the staging buffer and then submit the
+         * entire blob to the journal. We support passing it as sealed memfd in
+         * case it exceeds the datagram maximum.
+         * Sadly, `%r' is still no standardized format specifier, so we need
+         * the separate calls into log_append() here.
+         */
+
+        if (format && *format) {
+                log_appends(log, "MESSAGE=");
+                log_vappendf(log, format, args);
+                log_appends(log, "\n");
+        }
+
+        if (log->error)
+                r = error_trace(log->error);
+        else if (log->offset)
+                r = log_journal_send(log);
+        else
+                r = 0;
+
+        return error_trace(r);
+}
+
+/**
+ * log_vcommitf() - commit log message
+ * @log:                log to operate on
+ * @format:             log message format string
+ * @args:               arguments for format string
+ *
+ * This formats a log message and commits it. @format is used as format string
+ * for the log message, with @args filled in at the respective places.
+ *
+ * Any previously appended log fields are amended to the log message and
+ * submitted with it. In case the log output does not support structured
+ * logging, they will be lost, and only the message itself is submitted.
+ *
+ * If the pending log message is poisoned, or if it could not be submitted,
+ * an error will be returned. After this call returns (regardless whether it
+ * failed or not), the log context is ready to take the next log message.
+ *
+ * Return: 0 on success, LOG_E_TRUNCATED if the message was truncated by the
+ *         log channel, LOG_E_OVERSIZED if the message exceeded the maximum
+ *         size, and negative error code on failure.
+ */
+int log_vcommitf(Log *log, const char *format, va_list args) {
+        int r;
+
+        switch (log->mode) {
+        case LOG_MODE_STDERR:
+                r = log_commit_stderr(log, format, args);
+                break;
+        case LOG_MODE_JOURNAL:
+                r = log_commit_journal(log, format, args);
+                break;
+        default:
+                r = error_trace(log->error);
```

Comment:
Why this error? And in case `log->error` is 0, should we not be returning an error regardless in this case?

Reply:
The default is /dev/null, which is handled here. Why should we return an error?

Reply:
Make it an explicit case? Just to make it more obvious that this is expected.